### PR TITLE
MM-40813: Add config interface

### PR DESCRIPTION
**Pull request by agnivade**
We create the basic boilerplate to pass services
to each product. The basic idea is to have a map
containing service names and the services.

The server will pass all services to all products.
But it is on the product, to choose what services
they will actually consume.

Each product will cast the received service to
an interface with the methods that product requires.

https://mattermost.atlassian.net/browse/MM-40813
https://mattermost.atlassian.net/browse/MM-40816

```release-note
NONE
```

### CODE REVIEW

**Comment by agnivade** [comment]
Let me know how this approach looks. I want to implement the config service in this PR itself if the interface looks good to you.

File: app/channels.go [hunk 1]
```diff
@@ -65,17 +77,40 @@ type Channels struct {
...
 func init() {
-	RegisterProduct("channels", func(s *Server) (Product, error) {
-		return NewChannels(s)
+	RegisterProduct("channels", func(s *Server, services map[ServiceKey]interface{}) (Product, error) {
+		return NewChannels(s, services)
 	})
 }
 
-func NewChannels(s *Server) (*Channels, error) {
+func NewChannels(s *Server, services map[ServiceKey]interface{}) (*Channels, error) {
```
Comment (isacikgoz) [inline app/channels.go:85]: Not urgent but maybe we can define an interface for services eg. `GetName()` to match with the keys.

**Review by isacikgoz — COMMENTED** [review]
Looks good, I think we can go with this approach.

File: app/channels.go (hunk 1 above)
Comment (agnivade) [reply app/channels.go:85]: I like it :+1:

**Review by isacikgoz — APPROVED** [review]

**Pull request merged**